src/protocol/modulus_conversion/gen_random.rs

```diff
@@ -0,0 +1,166 @@
+        channel_identity: Identity,
+    ) -> (Replicated<F>, Replicated<F>, Replicated<F>) {
+        match channel_identity {
+            Identity::H1 => (
+                Replicated::new(F::from(u128::from(input.0)), F::ZERO),
+                Replicated::new(F::ZERO, F::from(u128::from(input.1))),
+                Replicated::new(F::ZERO, F::ZERO),
+            ),
+            Identity::H2 => (
+                Replicated::new(F::ZERO, F::ZERO),
+                Replicated::new(F::from(u128::from(input.0)), F::ZERO),
+                Replicated::new(F::ZERO, F::from(u128::from(input.1))),
+            ),
+            Identity::H3 => (
+                Replicated::new(F::ZERO, F::from(u128::from(input.1))),
+                Replicated::new(F::ZERO, F::ZERO),
+                Replicated::new(F::from(u128::from(input.0)), F::ZERO),
+            ),
+        }
+    }
+
+    async fn xor<F: Field, M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        a: Replicated<F>,
+        b: Replicated<F>,
+        ctx: &ProtocolContext<'_, G, S>,
+        step: S,
+        record_id: RecordId,
+    ) -> Result<Replicated<F>, BoxError> {
+        let result = ctx.multiply(record_id, step).await.execute(a, b).await?;
+
+        Ok(a + b - (result * F::from(2)))
+    }
+
+    #[allow(dead_code)]
+    pub async fn execute<F: Field, M: Mesh, G: Gateway<M, S>, S: Step + SpaceIndex>(
+        &self,
+        ctx: &ProtocolContext<'_, G, S>,
+        record_id: RecordId,
+        step1: S,
+        step2: S,
+    ) -> Result<Replicated<F>, BoxError> {
+        let (sh0, sh1, sh2) = Self::local_secret_share(self.input, ctx.identity);
+
+        let sh0_xor_sh1 = Self::xor(sh0, sh1, ctx, step1, record_id).await?;
+        Self::xor(sh0_xor_sh1, sh2, ctx, step2, record_id).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use proptest::prelude::Rng;
+    use tokio::try_join;
+
+    use crate::{
+        field::{Field, Fp31},
+        helpers::mock::{TestHelperGateway, TestMesh},
+        protocol::{
+            modulus_conversion::gen_random::{GenRandom, ModulusConversionStep, ReplicatedBinary},
+            QueryId, RecordId, SpaceIndex, Step,
+        },
+        test_fixture::{make_contexts, make_world, validate_and_reconstruct, TestWorld},
+    };
+
+    #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+    struct ModulusConversionTestStep {
+        bit_number: u8,
+        internal_step: ModulusConversionStep,
+    }
+
+    impl Step for ModulusConversionTestStep {}
+
+    impl SpaceIndex for ModulusConversionTestStep {
+        const MAX: usize = 512;
+
+        fn as_usize(&self) -> usize {
+            let b = self.bit_number as usize;
+            match self.internal_step {
+                ModulusConversionStep::Xor1 => b,
+                ModulusConversionStep::Xor2 => (256_usize + b),
+            }
+        }
+    }
+
+    #[tokio::test]
+    pub async fn gen_random() {
+        let mut rng = rand::thread_rng();
+
+        for i in 0_u8..40 {
```

Comment:
it may be good to test that we can run these conversions in parallel by spawning a future per record and collecting them at the end

Reply:
I agree. I would like to do that. I got a little stuck trying to figure out how to do so =(.

Can you please suggest some sample code that does this?

Reply:
I am looking at the latest rev and it seems it is already spawning them in parallel :)